### DISCUSSION TITLE
How can I write a library which has multiple files using `require`?

### DISCUSSION THREAD

**Discussion by jorolf** [cc-tweaked/CC-Tweaked#1176] [Q&A] [answered]
I'm currently working on a library which has multiple files that depend on each other. (not cyclically)

Imagine:
```
/lib
|---foo.lua
\---bar.lua
```

How can I consistently load `bar.lua` inside `foo.lua`? Using `require` means that I depend on `package.path` which is set depending on the main lua file:

```
main1.lua
lib/
|---examples/
    \---main3.lua
|---foo.lua
|---bar.lua
\---main2.lua
```

For `main1.lua` I would have to require `lib.bar`
For `main2.lua` I would have to require just `bar`
And for `main3.lua` it's impossible because `../bar` doesn't work

Note that I don't mean `main -require-> bar` but `foo -require-> bar`

**Comment by SquidDev** [comment, marked as answer]
I'd set `package.path` inside your entrypoints, so that `lib` is always on the path. This way both your entrypoints and other libraries only need to do `require "foo"`/`require "bar"`.

So, in the case of `main1.lua`, something like 

```lua
package.path = "lib/?.lua:" .. package.path 

local foo = require "foo"
```

**Reply by jorolf** [reply]
This is probably the simplest solution. However I don't like how it puts the burdon on the library user and not the implementor.... Do you know if there's another way?

**Reply by SquidDev** [reply]
It only really puts the burden on the library user if they want to do something different with folder layout - if things are in the same directory, the user doesn't have to anything extra.

There's something nasty one could do with module names, but I don't think it's the right solution - I'll write it up as a separate answer.

**Reply by jorolf** [reply]
Thanks for the suggestions. I've settled for the simpler variant. Your example contains an error though: It should be `"lib/?.lua;"` instead of `"lib/?.lua:"`

> It only really puts the burden on the library user if they want to do something different with folder layout

I like to organize my files into modules or packages instead of just placing them all in the same directory/file

**Comment by SquidDev** [comment]
The current module name (`foo`, `lib.foo`) gets passed to the required module, so you _could_ use that to implement a sort-of-relative import. It's not very nice though:

```lua
local function get_prefix(str)
  local pos
  while true do
    local next = str:find(".", pos, true)
    if not next then break end
    pos = next + 1
  end

  if pos then return str:sub(1, pos - 1) else return "" end
end
local prefix = get_prefix(...) -- Will be something like "lib." or "".

local bar = require(prefix .. "bar") -- So will be "lib.bar" or "bar"
```
